// harmony-marketplace-sdk/lib.esm/bridge/bridgedHrc721Token/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { EthBaseContract } from '../../contracts';
import ABI from './abi';
export class BridgedHRC721Token extends EthBaseContract {
    constructor(address, signer) {
        super(address, ABI, signer);
    }
    totalSupply() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('totalSupply');
        });
    }
    balanceOf(address) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('balanceOf', [address]);
        });
    }
    approve(spender, tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('approve', [spender, tokenId], txOptions);
        });
    }
    transferFrom(from, to, tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('transferFrom', [from, to, tokenId], txOptions);
        });
    }
    symbol() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('symbol');
        });
    }
    name() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('name');
        });
    }
    tokenURI(tokenId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('tokenURI', [tokenId]);
        });
    }
}
