// harmony-marketplace-sdk/lib.esm/bridge/hrc20HmyManager/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { BaseContract } from '../../contracts';
import ABI from './abi';
export class HRC20HmyManager extends BaseContract {
    constructor(address, provider, options) {
        super(address, ABI, provider, options);
    }
    lockToken(hmyTokenAddr, amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockToken', [hmyTokenAddr, amount, recipient], txOptions);
        });
    }
    lockTokenFor(hmyTokenAddr, userAddr, amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockTokenFor', [hmyTokenAddr, userAddr, amount, recipient], txOptions);
        });
    }
    unlockToken(hmyTokenAddr, amount, recipient, receiptId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('unlockToken', [hmyTokenAddr, amount, recipient, receiptId], txOptions);
        });
    }
    lockOne(amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockOne', [amount, recipient], txOptions);
        });
    }
    unlockOne(amount, recipient, receiptId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('unlockOne', [amount, recipient, receiptId], txOptions);
        });
    }
}
