// harmony-marketplace-sdk/lib/bridge/bridgeToken.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.BridgeToken = void 0;
var abstract_signer_1 = require("@ethersproject/abstract-signer");
var constants_1 = require("../constants");
var wallets_1 = require("../wallets");
var BridgeToken = (function () {
    function BridgeToken(hmyOwnerWallet, ethOwnerWallet, network) {
        if (network === void 0) { network = constants_1.NetworkInfo.MAINNET; }
        this.network = network;
        this.isMainnet = this.network === constants_1.NetworkInfo.MAINNET;
        if (this.isMainnet) {
            this.ethMasterWallet = new abstract_signer_1.VoidSigner(constants_1.MAINNET_MULTISIG_WALLET);
            this.hmyMasterWallet = new wallets_1.Key(constants_1.HARMONY_RPC_SHARD_0_URL);
        }
        else {
            this.ethMasterWallet = new abstract_signer_1.VoidSigner(constants_1.DEVNET_MULTISIG_WALLET);
            this.hmyMasterWallet = new wallets_1.Key(constants_1.HARMONY_RPC_SHARD_0_DEVNET_URL);
        }
        if (!ethOwnerWallet.provider) {
            throw new Error('Ethereum signer should have a rpc provider set.');
        }
        this.ethMasterWallet = this.ethMasterWallet.connect(ethOwnerWallet.provider);
        this.hmyMasterWallet.setSigner(constants_1.MAINNET_MULTISIG_WALLET);
        this.hmyOwnerWallet = hmyOwnerWallet;
        this.ethOwnerWallet = ethOwnerWallet;
    }
    BridgeToken.prototype.sendToken = function (type, sender, recipient, token, tokenInfo, txOptions) {
        if (txOptions === void 0) { txOptions = constants_1.DEFAULT_TX_OPTIONS; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (type === constants_1.BridgeType.ETH_TO_HMY) {
                    return [2, this.ethToHmy(sender, recipient, token, tokenInfo)];
                }
                else {
                    return [2, this.hmyToEth(sender, recipient, token, tokenInfo, txOptions)];
                }
                return [2];
            });
        });
    };
    return BridgeToken;
}());
exports.BridgeToken = BridgeToken;
