// harmony-marketplace-sdk/lib.esnext/bridge/hrc20EthManager/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRC20EthManager = void 0;
const contracts_1 = require("../../contracts");
const abi_1 = __importDefault(require("./abi"));
class HRC20EthManager extends contracts_1.EthBaseContract {
    constructor(address, signer) {
        super(address, abi_1.default, signer);
    }
    addToken(tokenManager, hmyTokenAddr, name, symbol, decimals, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('addToken', [tokenManager, hmyTokenAddr, name, symbol, decimals], txOptions);
        });
    }
    removeToken(tokenManager, hmyTokenAddr, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('removeToken', [tokenManager, hmyTokenAddr], txOptions);
        });
    }
    burnToken(ethToken, amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('burnToken', [ethToken, amount, recipient], txOptions);
        });
    }
    mintToken(ethToken, amount, recipient, receiptId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('mintToken', [ethToken, amount, recipient, receiptId], txOptions);
        });
    }
    mappings(hrc20Addr) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('mappings', [hrc20Addr]);
        });
    }
}
exports.HRC20EthManager = HRC20EthManager;
