// harmony-marketplace-sdk/lib.esnext/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TX_OPTIONS = exports.DEVNET_HRC1155_CONTRACTS_ADDRESSES = exports.DEVNET_HRC721_CONTRACTS_ADDRESSES = exports.DEVNET_HRC20_CONTRACTS_ADDRESSES = exports.DEVNET_MULTISIG_WALLET = exports.MAINNET_HRC1155_CONTRACTS_ADDRESSES = exports.MAINNET_HRC721_CONTRACTS_ADDRESSES = exports.MAINNET_HRC20_CONTRACTS_ADDRESSES = exports.MAINNET_MULTISIG_WALLET = exports.BridgeType = exports.TokenType = exports.NetworkInfo = exports.HARMONY_SHARDS = exports.HARMONY_RPC_SHARD_0_DEVNET = exports.HARMONY_RPC_SHARD_3 = exports.HARMONY_RPC_SHARD_2 = exports.HARMONY_RPC_SHARD_1 = exports.HARMONY_RPC_SHARD_0 = exports.HARMONY_RPC_DEVNET_WS = exports.HARMONY_RPC_WS = exports.HARMONY_RPC_SHARD_0_DEVNET_URL = exports.HARMONY_RPC_SHARD_3_URL = exports.HARMONY_RPC_SHARD_2_URL = exports.HARMONY_RPC_SHARD_1_URL = exports.HARMONY_RPC_SHARD_0_URL = exports.DEFAULT_GAS_LIMIT = exports.DEFAULT_GAS_PRICE = exports.AddressZero = void 0;
const utils_1 = require("@harmony-js/utils");
const interfaces_1 = require("./interfaces");
exports.AddressZero = '0x0000000000000000000000000000000000000000';
exports.DEFAULT_GAS_PRICE = '1000000000';
exports.DEFAULT_GAS_LIMIT = '3500000';
exports.HARMONY_RPC_SHARD_0_URL = 'https://api.harmony.one';
exports.HARMONY_RPC_SHARD_1_URL = 'https://s1.api.harmony.one';
exports.HARMONY_RPC_SHARD_2_URL = 'https://s2.api.harmony.one';
exports.HARMONY_RPC_SHARD_3_URL = 'https://s3.api.harmony.one';
exports.HARMONY_RPC_SHARD_0_DEVNET_URL = 'https://api.s0.ps.hmny.io';
exports.HARMONY_RPC_WS = 'wss://ws.s0.t.hmny.io';
exports.HARMONY_RPC_DEVNET_WS = 'wss://ws.s0.ps.hmny.io';
exports.HARMONY_RPC_SHARD_0 = {
    url: exports.HARMONY_RPC_SHARD_0_URL,
    chainId: utils_1.ChainID.HmyMainnet,
    chainType: utils_1.ChainType.Harmony,
};
exports.HARMONY_RPC_SHARD_1 = {
    url: exports.HARMONY_RPC_SHARD_1_URL,
    chainId: utils_1.ChainID.HmyMainnet,
    chainType: utils_1.ChainType.Harmony,
};
exports.HARMONY_RPC_SHARD_2 = {
    url: exports.HARMONY_RPC_SHARD_2_URL,
    chainId: utils_1.ChainID.HmyMainnet,
    chainType: utils_1.ChainType.Harmony,
};
exports.HARMONY_RPC_SHARD_3 = {
    url: exports.HARMONY_RPC_SHARD_3_URL,
    chainId: utils_1.ChainID.HmyMainnet,
    chainType: utils_1.ChainType.Harmony,
};
exports.HARMONY_RPC_SHARD_0_DEVNET = {
    url: exports.HARMONY_RPC_SHARD_0_DEVNET_URL,
    chainId: 4,
    chainType: utils_1.ChainType.Harmony,
};
exports.HARMONY_SHARDS = {
    [interfaces_1.HarmonyShards.SHARD_0]: exports.HARMONY_RPC_SHARD_0,
    [interfaces_1.HarmonyShards.SHARD_1]: exports.HARMONY_RPC_SHARD_1,
    [interfaces_1.HarmonyShards.SHARD_2]: exports.HARMONY_RPC_SHARD_2,
    [interfaces_1.HarmonyShards.SHARD_3]: exports.HARMONY_RPC_SHARD_3,
    [interfaces_1.HarmonyShards.SHARD_0_DEVNET]: exports.HARMONY_RPC_SHARD_0_DEVNET,
};
var NetworkInfo;
(function (NetworkInfo) {
    NetworkInfo[NetworkInfo["MAINNET"] = 0] = "MAINNET";
    NetworkInfo[NetworkInfo["DEVNET"] = 1] = "DEVNET";
})(NetworkInfo = exports.NetworkInfo || (exports.NetworkInfo = {}));
var TokenType;
(function (TokenType) {
    TokenType[TokenType["HRC20"] = 0] = "HRC20";
    TokenType[TokenType["HRC721"] = 1] = "HRC721";
    TokenType[TokenType["HRC1155"] = 2] = "HRC1155";
})(TokenType = exports.TokenType || (exports.TokenType = {}));
var BridgeType;
(function (BridgeType) {
    BridgeType[BridgeType["HMY_TO_ETH"] = 0] = "HMY_TO_ETH";
    BridgeType[BridgeType["ETH_TO_HMY"] = 1] = "ETH_TO_HMY";
})(BridgeType = exports.BridgeType || (exports.BridgeType = {}));
exports.MAINNET_MULTISIG_WALLET = '0x715CdDa5e9Ad30A0cEd14940F9997EE611496De6';
exports.MAINNET_HRC20_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0x2912885736Ce25E437c0113200254140a709a58d',
    hmyManagerAddress: '0xe0c1267f1c63e83472d2b3e25d970740940d752b',
    tokenManagerAddress: '0x92591545c6462ad0751c890D15DF5F47846fF4Bf',
};
exports.MAINNET_HRC721_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0x426A61A2127fDD1318Ec0EdCe02474f382FdAd30',
    hmyManagerAddress: '0xbaf4d51738a42b976c3558b5f983cf4721451499',
    tokenManagerAddress: '0xF837fe0Eba85bE14446E546115ef20891E357D2B',
};
exports.MAINNET_HRC1155_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0x478279c5A0beb8401De1b4EaCB4863a243a8e3A3',
    hmyManagerAddress: '0x4f9b3defb4f61227a7574f2a7adfe2841e1ae20e',
    tokenManagerAddress: '0x94da065b27f4a61d6595c2ebb541bb7bd11b6266',
};
exports.DEVNET_MULTISIG_WALLET = '0xC12208C19f895df6BD589fD4F574CCf53992B9F5';
exports.DEVNET_HRC20_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0xb6bE28Dc813D52A4740c3FA9C0c04e6542cce48d',
    hmyManagerAddress: '0xe494ebf0290833ee655e4edecca95ef82ae60c4e',
    tokenManagerAddress: '0x4a9832Aa2b22417Be739b15391Ec386403461f4b',
};
exports.DEVNET_HRC721_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0x5c8a5cC4b46C8cdD2C50b3357984086814956DA7',
    hmyManagerAddress: '0x2150cbd4a66f8922a21eec68c1c9cafdaca5eddc',
    tokenManagerAddress: '0xCda2C033FDD220D987555e9F4934E6E69C5BC4Bd',
};
exports.DEVNET_HRC1155_CONTRACTS_ADDRESSES = {
    ethManagerAddress: '0x30Ec2A1002D1b099537d243b8A753A328A8f5ff1',
    hmyManagerAddress: '0x7bf2fbfd386e61776085bd4b69821145fb8b059f',
    tokenManagerAddress: '0x61BC4C52cE08A4BF5A3d5CBcF5d1664842c64f53',
};
exports.DEFAULT_TX_OPTIONS = {
    gasPrice: new utils_1.Unit('30').asGwei().toWei(),
    gasLimit: 3500000,
};
