// harmony-marketplace-sdk/lib.esm/contracts/hrc1155.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AddressZero } from '../constants';
import * as Utils from '../utils';
import { ContractError } from './baseContract';
import { BaseToken } from './baseToken';
export class HRC1155 extends BaseToken {
    balanceOf(address, id, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getBalance(address, id, txOptions);
        });
    }
    balanceOfBatch(accounts, ids, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (accounts.length !== ids.length) {
                throw new ContractError('Accounts and ids must have the same length', 'balanceOfBatch');
            }
            return this.call('balanceOfBatch', [accounts, ids], txOptions);
        });
    }
    safeTransferFrom(from, to, id, amount, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (to === AddressZero) {
                throw new ContractError(`The to cannot be the ${AddressZero}`, 'safeTransferFrom');
            }
            return this.send('safeTransferFrom', [from, to, id, amount, data], txOptions);
        });
    }
    safeBatchTransferFrom(from, to, ids, amounts, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (amounts.length !== ids.length) {
                throw new ContractError('amounts and ids must have the same length', 'safeBatchTransferFrom');
            }
            return this.send('safeBatchTransferFrom', [from, to, ids, amounts, data], txOptions);
        });
    }
    setApprovalForAll(addressOperator, approved, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!addressOperator) {
                throw new Error('You must provide an addressOperator');
            }
            return this.send('setApprovalForAll', [addressOperator, approved], txOptions);
        });
    }
    isApprovedForAll(owner, operator, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!owner || owner === AddressZero) {
                throw new ContractError('Invalid owner provided', 'isApprovedForAll');
            }
            if (!operator || operator === AddressZero) {
                throw new ContractError('Invalid operator provided', 'isApprovedForAll');
            }
            return this.call('isApprovedForAll', [owner, operator], txOptions);
        });
    }
    owner(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const address = yield this.call('owner', [], txOptions);
            return this.sanitizeAddress(address);
        });
    }
    tokenURIPrefix(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('tokenURIPrefix', [], txOptions);
        });
    }
    contractURI(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('contractURI', [], txOptions);
        });
    }
    totalSupply(id, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('totalSupply', [id], txOptions);
        });
    }
    tokenURI(tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!Utils.isBNish(tokenId)) {
                throw new ContractError('You must provide a tokenId', 'uri');
            }
            return this.call('uri', [tokenId], txOptions);
        });
    }
    symbol(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('symbol', [], txOptions);
        });
    }
    name(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('name', [], txOptions);
        });
    }
    mint(account, tokenId, amount, txOptions) {
        return this.send('mint', [account, tokenId, amount, []], txOptions);
    }
    mintBatch(account, tokenIds, amounts, txOptions) {
        return this.send('mintBatch', [account, tokenIds, amounts, []], txOptions);
    }
}
