// harmony-marketplace-sdk/lib.esm/contracts/hrc721.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AddressZero } from '../constants';
import * as Utils from '../utils';
import { ContractError } from './baseContract';
import { BaseToken } from './baseToken';
export class HRC721 extends BaseToken {
    balanceOf(address, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.getBalance(address, undefined, txOptions);
        });
    }
    ownerOf(tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!Utils.isBNish(tokenId)) {
                throw new ContractError('You must provide a tokenId', 'ownerOf');
            }
            const address = yield this.call('ownerOf', [tokenId], txOptions);
            return this.sanitizeAddress(address);
        });
    }
    safeTransferFrom(from, to, tokenId, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = [from, to, tokenId];
            if (data) {
                args.push(data);
            }
            return this.send('safeTransferFrom', args, txOptions);
        });
    }
    transferFrom(from, to, tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('transferFrom', [from, to, tokenId], txOptions);
        });
    }
    approve(to, tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('approve', [to, tokenId], txOptions);
        });
    }
    getApproved(tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!Utils.isBNish(tokenId)) {
                throw new ContractError('You must provide a tokenId', 'getApproved');
            }
            const address = yield this.call('getApproved', [tokenId], txOptions);
            return this.sanitizeAddress(address);
        });
    }
    setApprovalForAll(addressOperator, approved, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!addressOperator) {
                throw new Error('You must provide an addressOperator');
            }
            return this.send('setApprovalForAll', [addressOperator, approved], txOptions);
        });
    }
    isApprovedForAll(owner, operator, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!owner || owner === AddressZero) {
                throw new ContractError('Invalid owner provided', 'isApprovedForAll');
            }
            if (!operator || operator === AddressZero) {
                throw new ContractError('Invalid operator provided', 'isApprovedForAll');
            }
            return this.call('isApprovedForAll', [owner, operator], txOptions);
        });
    }
    totalSupply(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('totalSupply', [], txOptions);
        });
    }
    tokenURI(tokenId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!Utils.isBNish(tokenId)) {
                throw new ContractError('You must provide a tokenId', 'tokenURI');
            }
            return this.call('tokenURI', [tokenId], txOptions);
        });
    }
    symbol(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('symbol', [], txOptions);
        });
    }
    name(txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.call('name', [], txOptions);
        });
    }
    mint(account, tokenId, txOptions) {
        return this.send('mint', [account, tokenId], txOptions);
    }
    safeMint(to, tokenId, txOptions) {
        return this.send('safeMint', [to, tokenId], txOptions);
    }
    burn(tokenId, txOptions) {
        return this.send('burn', [tokenId], txOptions);
    }
}
