// harmony-marketplace-sdk/lib.esnext/bridge/hrc1155EthManager/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRC1155EthManager = void 0;
const contracts_1 = require("../../contracts");
const abi_1 = __importDefault(require("./abi"));
class HRC1155EthManager extends contracts_1.EthBaseContract {
    constructor(address, signer) {
        super(address, abi_1.default, signer);
    }
    addToken(tokenManager, ethTokenAddr, name, symbol, tokenURI, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('addToken', [tokenManager, ethTokenAddr, name, symbol, tokenURI], txOptions);
        });
    }
    removeToken(tokenManager, ethTokenAddr, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('removeToken', [tokenManager, ethTokenAddr], txOptions);
        });
    }
    burnToken(oneToken, tokenId, recipient, amount, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('burnToken', [oneToken, tokenId, recipient, amount], txOptions);
        });
    }
    burnTokens(oneToken, tokenIds, recipient, amounts, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('burnTokens', [oneToken, tokenIds, recipient, amounts], txOptions);
        });
    }
    mintToken(oneToken, tokenId, recipient, receiptId, amount, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('mintToken', [oneToken, tokenId, recipient, receiptId, amount, data], txOptions);
        });
    }
    mintTokens(oneToken, tokenIds, amounts, recipient, receiptId, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('mintTokens', [oneToken, tokenIds, recipient, receiptId, amounts, data], txOptions);
        });
    }
    mappings(hrc1155Addr) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('mappings', [hrc1155Addr]);
        });
    }
}
exports.HRC1155EthManager = HRC1155EthManager;
