// harmony-marketplace-sdk/lib/bridge/bridgedHrc1155Token/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.BridgedHRC1155Token = void 0;
var constants_1 = require("../../constants");
var contracts_1 = require("../../contracts");
var abi_1 = __importDefault(require("./abi"));
var BridgedHRC1155Token = (function (_super) {
    __extends(BridgedHRC1155Token, _super);
    function BridgedHRC1155Token(address, signer) {
        return _super.call(this, address, abi_1["default"], signer) || this;
    }
    BridgedHRC1155Token.prototype.balanceOf = function (address, id) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.read('balanceOf', [address, id])];
            });
        });
    };
    BridgedHRC1155Token.prototype.balanceOfBatch = function (accounts, ids) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (accounts.length !== ids.length) {
                    throw new contracts_1.ContractError('Accounts and ids must have the same length', 'balanceOfBatch');
                }
                return [2, this.read('balanceOfBatch', [accounts, ids])];
            });
        });
    };
    BridgedHRC1155Token.prototype.safeTransferFrom = function (from, to, id, amount, data, txOptions) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (to === constants_1.AddressZero) {
                    throw new contracts_1.ContractError("The to cannot be the ".concat(constants_1.AddressZero), 'safeTransferFrom');
                }
                return [2, this.write('safeTransferFrom', [from, to, id, amount, data], txOptions)];
            });
        });
    };
    BridgedHRC1155Token.prototype.safeBatchTransferFrom = function (from, to, ids, amounts, data, txOptions) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (amounts.length !== ids.length) {
                    throw new contracts_1.ContractError('amounts and ids must have the same length', 'safeBatchTransferFrom');
                }
                return [2, this.write('safeBatchTransferFrom', [from, to, ids, amounts, data], txOptions)];
            });
        });
    };
    BridgedHRC1155Token.prototype.setApprovalForAll = function (addressOperator, approved, txOptions) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!addressOperator) {
                    throw new Error('You must provide an addressOperator');
                }
                return [2, this.write('setApprovalForAll', [addressOperator, approved], txOptions)];
            });
        });
    };
    BridgedHRC1155Token.prototype.isApprovedForAll = function (owner, operator) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!owner || owner === constants_1.AddressZero) {
                    throw new contracts_1.ContractError('Invalid owner provided', 'isApprovedForAll');
                }
                if (!operator || operator === constants_1.AddressZero) {
                    throw new contracts_1.ContractError('Invalid operator provided', 'isApprovedForAll');
                }
                return [2, this.read('isApprovedForAll', [owner, operator])];
            });
        });
    };
    BridgedHRC1155Token.prototype.owner = function () {
        return __awaiter(this, void 0, void 0, function () {
            var address;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.read('owner', [])];
                    case 1:
                        address = _a.sent();
                        return [2, address.toLowerCase()];
                }
            });
        });
    };
    BridgedHRC1155Token.prototype.tokenURIPrefix = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.read('tokenURIPrefix', [])];
            });
        });
    };
    BridgedHRC1155Token.prototype.contractURI = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.read('contractURI', [])];
            });
        });
    };
    BridgedHRC1155Token.prototype.totalSupply = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.read('totalSupply', [id])];
            });
        });
    };
    return BridgedHRC1155Token;
}(contracts_1.EthBaseContract));
exports.BridgedHRC1155Token = BridgedHRC1155Token;
