// harmony-marketplace-sdk/lib.esm/bridge/hrc721TokenManager/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { EthBaseContract } from '../../contracts';
import ABI from './abi';
export class HRC721TokenManager extends EthBaseContract {
    constructor(address, signer) {
        super(address, ABI, signer);
    }
    rely(address, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('rely', [address], txOptions);
        });
    }
    deny(address, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('deny', [address], txOptions);
        });
    }
    addToken(ethTokenAddr, name, symbol, tokenURI, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('addToken', [ethTokenAddr, name, symbol, tokenURI], txOptions);
        });
    }
    registerToken(ethTokenAddr, oneTokenAddr, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('registerToken', [ethTokenAddr, oneTokenAddr], txOptions);
        });
    }
    removeToken(ethTokenAddr, supply, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('removeToken', [ethTokenAddr, supply], txOptions);
        });
    }
}
