// harmony-marketplace-sdk/lib.esnext/bridge/bridgedHrc1155Token/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgedHRC1155Token = void 0;
const constants_1 = require("../../constants");
const contracts_1 = require("../../contracts");
const abi_1 = __importDefault(require("./abi"));
class BridgedHRC1155Token extends contracts_1.EthBaseContract {
    constructor(address, signer) {
        super(address, abi_1.default, signer);
    }
    balanceOf(address, id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('balanceOf', [address, id]);
        });
    }
    balanceOfBatch(accounts, ids) {
        return __awaiter(this, void 0, void 0, function* () {
            if (accounts.length !== ids.length) {
                throw new contracts_1.ContractError('Accounts and ids must have the same length', 'balanceOfBatch');
            }
            return this.read('balanceOfBatch', [accounts, ids]);
        });
    }
    safeTransferFrom(from, to, id, amount, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (to === constants_1.AddressZero) {
                throw new contracts_1.ContractError(`The to cannot be the ${constants_1.AddressZero}`, 'safeTransferFrom');
            }
            return this.write('safeTransferFrom', [from, to, id, amount, data], txOptions);
        });
    }
    safeBatchTransferFrom(from, to, ids, amounts, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (amounts.length !== ids.length) {
                throw new contracts_1.ContractError('amounts and ids must have the same length', 'safeBatchTransferFrom');
            }
            return this.write('safeBatchTransferFrom', [from, to, ids, amounts, data], txOptions);
        });
    }
    setApprovalForAll(addressOperator, approved, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!addressOperator) {
                throw new Error('You must provide an addressOperator');
            }
            return this.write('setApprovalForAll', [addressOperator, approved], txOptions);
        });
    }
    isApprovedForAll(owner, operator) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!owner || owner === constants_1.AddressZero) {
                throw new contracts_1.ContractError('Invalid owner provided', 'isApprovedForAll');
            }
            if (!operator || operator === constants_1.AddressZero) {
                throw new contracts_1.ContractError('Invalid operator provided', 'isApprovedForAll');
            }
            return this.read('isApprovedForAll', [owner, operator]);
        });
    }
    owner() {
        return __awaiter(this, void 0, void 0, function* () {
            const address = yield this.read('owner', []);
            return address.toLowerCase();
        });
    }
    tokenURIPrefix() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('tokenURIPrefix', []);
        });
    }
    contractURI() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('contractURI', []);
        });
    }
    totalSupply(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('totalSupply', [id]);
        });
    }
}
exports.BridgedHRC1155Token = BridgedHRC1155Token;
