// harmony-marketplace-sdk/lib.esm/bridge/bridgedHrc1155Token/index.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AddressZero } from '../../constants';
import { ContractError, EthBaseContract } from '../../contracts';
import ABI from './abi';
export class BridgedHRC1155Token extends EthBaseContract {
    constructor(address, signer) {
        super(address, ABI, signer);
    }
    balanceOf(address, id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('balanceOf', [address, id]);
        });
    }
    balanceOfBatch(accounts, ids) {
        return __awaiter(this, void 0, void 0, function* () {
            if (accounts.length !== ids.length) {
                throw new ContractError('Accounts and ids must have the same length', 'balanceOfBatch');
            }
            return this.read('balanceOfBatch', [accounts, ids]);
        });
    }
    safeTransferFrom(from, to, id, amount, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (to === AddressZero) {
                throw new ContractError(`The to cannot be the ${AddressZero}`, 'safeTransferFrom');
            }
            return this.write('safeTransferFrom', [from, to, id, amount, data], txOptions);
        });
    }
    safeBatchTransferFrom(from, to, ids, amounts, data, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (amounts.length !== ids.length) {
                throw new ContractError('amounts and ids must have the same length', 'safeBatchTransferFrom');
            }
            return this.write('safeBatchTransferFrom', [from, to, ids, amounts, data], txOptions);
        });
    }
    setApprovalForAll(addressOperator, approved, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!addressOperator) {
                throw new Error('You must provide an addressOperator');
            }
            return this.write('setApprovalForAll', [addressOperator, approved], txOptions);
        });
    }
    isApprovedForAll(owner, operator) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!owner || owner === AddressZero) {
                throw new ContractError('Invalid owner provided', 'isApprovedForAll');
            }
            if (!operator || operator === AddressZero) {
                throw new ContractError('Invalid operator provided', 'isApprovedForAll');
            }
            return this.read('isApprovedForAll', [owner, operator]);
        });
    }
    owner() {
        return __awaiter(this, void 0, void 0, function* () {
            const address = yield this.read('owner', []);
            return address.toLowerCase();
        });
    }
    tokenURIPrefix() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('tokenURIPrefix', []);
        });
    }
    contractURI() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('contractURI', []);
        });
    }
    totalSupply(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.read('totalSupply', [id]);
        });
    }
}
