// harmony-marketplace-sdk/lib.esnext/bridge/bridgeToken.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeToken = void 0;
const abstract_signer_1 = require("@ethersproject/abstract-signer");
const constants_1 = require("../constants");
const wallets_1 = require("../wallets");
class BridgeToken {
    constructor(hmyOwnerWallet, ethOwnerWallet, network = constants_1.NetworkInfo.MAINNET) {
        this.network = network;
        this.isMainnet = this.network === constants_1.NetworkInfo.MAINNET;
        if (this.isMainnet) {
            this.ethMasterWallet = new abstract_signer_1.VoidSigner(constants_1.MAINNET_MULTISIG_WALLET);
            this.hmyMasterWallet = new wallets_1.Key(constants_1.HARMONY_RPC_SHARD_0_URL);
        }
        else {
            this.ethMasterWallet = new abstract_signer_1.VoidSigner(constants_1.DEVNET_MULTISIG_WALLET);
            this.hmyMasterWallet = new wallets_1.Key(constants_1.HARMONY_RPC_SHARD_0_DEVNET_URL);
        }
        if (!ethOwnerWallet.provider) {
            throw new Error('Ethereum signer should have a rpc provider set.');
        }
        this.ethMasterWallet = this.ethMasterWallet.connect(ethOwnerWallet.provider);
        this.hmyMasterWallet.setSigner(constants_1.MAINNET_MULTISIG_WALLET);
        this.hmyOwnerWallet = hmyOwnerWallet;
        this.ethOwnerWallet = ethOwnerWallet;
    }
    sendToken(type, sender, recipient, token, tokenInfo, txOptions = constants_1.DEFAULT_TX_OPTIONS) {
        return __awaiter(this, void 0, void 0, function* () {
            if (type === constants_1.BridgeType.ETH_TO_HMY) {
                return this.ethToHmy(sender, recipient, token, tokenInfo);
            }
            else {
                return this.hmyToEth(sender, recipient, token, tokenInfo, txOptions);
            }
        });
    }
}
exports.BridgeToken = BridgeToken;
