// harmony-marketplace-sdk/lib.esnext/bridge/hrc20TokenManager/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRC20TokenManager = void 0;
const contracts_1 = require("../../contracts");
const abi_1 = __importDefault(require("./abi"));
class HRC20TokenManager extends contracts_1.EthBaseContract {
    constructor(address, signer) {
        super(address, abi_1.default, signer);
    }
    rely(address, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('rely', [address], txOptions);
        });
    }
    deny(address, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('deny', [address], txOptions);
        });
    }
    addToken(ethTokenAddr, name, symbol, decimals, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('addToken', [ethTokenAddr, name, symbol, decimals], txOptions);
        });
    }
    registerToken(ethTokenAddr, oneTokenAddr, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('registerToken', [ethTokenAddr, oneTokenAddr], txOptions);
        });
    }
    removeToken(ethTokenAddr, supply, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.write('removeToken', [ethTokenAddr, supply], txOptions);
        });
    }
}
exports.HRC20TokenManager = HRC20TokenManager;
