// harmony-marketplace-sdk/lib/wallets/hd-key.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.HDKey = void 0;
var account_1 = require("@harmony-js/account");
var transaction_1 = require("@harmony-js/transaction");
var utils_1 = require("@harmony-js/utils");
var key_1 = require("./key");
var HDKey = (function (_super) {
    __extends(HDKey, _super);
    function HDKey(url, options, chainId, chainType) {
        var _this = _super.call(this, url, chainId, chainType) || this;
        var _a = options.mnemonic, mnemonic = _a === void 0 ? account_1.Wallet.generateMnemonic() : _a, _b = options.index, index = _b === void 0 ? 0 : _b, _c = options.numberOfAddresses, numberOfAddresses = _c === void 0 ? 1 : _c, _d = options.shardId, shardId = _d === void 0 ? 0 : _d, _e = options.gasLimit, gasLimit = _e === void 0 ? '1000000' : _e, _f = options.gasPrice, gasPrice = _f === void 0 ? '2000000000' : _f;
        _this.hdNode = new account_1.HDNode(_this.messenger.provider.url, mnemonic, index, numberOfAddresses, shardId, utils_1.ChainType.Harmony, chainId, gasLimit, gasPrice);
        return _this;
    }
    HDKey.prototype.signTransaction = function (transaction) {
        return __awaiter(this, void 0, void 0, function () {
            var rawTransaction, params, tx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.hdNode.signTransaction(transaction.txParams)];
                    case 1:
                        rawTransaction = _a.sent();
                        params = (0, transaction_1.recover)(rawTransaction);
                        tx = new transaction_1.Transaction({}, this.messenger);
                        tx.setParams(params);
                        return [2, tx];
                }
            });
        });
    };
    HDKey.prototype.setSigner = function (address) {
        this.hdNode.setSigner(address);
    };
    HDKey.prototype.addByPrivateKey = function (privateKey) {
        this.hdNode.addByPrivateKey(privateKey);
        return account_1.Account.add(privateKey);
    };
    HDKey.prototype.getAccount = function (address) {
        var addresses = this.hdNode.getAccounts();
        if (addresses.includes(address)) {
            return account_1.Account.add(this.hdNode.getPrivateKey(address));
        }
        return;
    };
    HDKey.prototype.setMessenger = function (messenger) {
        _super.prototype.setMessenger.call(this, messenger);
        this.hdNode.setProvider(this.messenger.provider);
    };
    HDKey.prototype.getAddress = function (idx) {
        return this.hdNode.getAddress(idx);
    };
    HDKey.prototype.getAddresses = function () {
        return this.hdNode.getAddresses();
    };
    return HDKey;
}(key_1.Key));
exports.HDKey = HDKey;
