// harmony-marketplace-sdk/lib.esnext/bridge/hrc20HmyManager/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRC20HmyManager = void 0;
const contracts_1 = require("../../contracts");
const abi_1 = __importDefault(require("./abi"));
class HRC20HmyManager extends contracts_1.BaseContract {
    constructor(address, provider, options) {
        super(address, abi_1.default, provider, options);
    }
    lockToken(hmyTokenAddr, amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockToken', [hmyTokenAddr, amount, recipient], txOptions);
        });
    }
    lockTokenFor(hmyTokenAddr, userAddr, amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockTokenFor', [hmyTokenAddr, userAddr, amount, recipient], txOptions);
        });
    }
    unlockToken(hmyTokenAddr, amount, recipient, receiptId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('unlockToken', [hmyTokenAddr, amount, recipient, receiptId], txOptions);
        });
    }
    lockOne(amount, recipient, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('lockOne', [amount, recipient], txOptions);
        });
    }
    unlockOne(amount, recipient, receiptId, txOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.send('unlockOne', [amount, recipient, receiptId], txOptions);
        });
    }
}
exports.HRC20HmyManager = HRC20HmyManager;
